ibus: 1.5.16 -> 1.5.17, fix

lukateras opened the pull request:
###### Motivation for this change


Fixes #30265, fixes #31454. `ibus-setup` is not broken anymore.

* Update cldrEmojiAnnotation: 31.0.1_1 -> 31.90.0_1
* Build from Git so that we don't have to patch `./configure` scripts when we could just generate them correctly
* Drop patch that removes IBUS_* environment variables from ibus-setup wrapper: they are actually set correctly, and we don't set them which breaks `ibus-setup`
* Add withWayland option (building with Wayland requires libxcbcommon package, previously wayland was included into buildInputs w/o libxcbcommon and respective configure flag)
* Add license (LGPLv2.1)

Relies on two other commits in this branch that add `GI_TYPELIB_PATH` discovery and fix `gconf` Python scripts (fixes #31083).

/cc @edolstra @globin

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

lukateras committed:
gconf: add python to dependencies, fixes #31083

lukateras committed:
ibus: 1.5.16 -> 1.5.17, fix, back to single output

This fixes ibus-setup.

* Update cldrEmojiAnnotation: 31.0.1_1 -> 31.90.0_1
* Move ibus-setup back to out, and use a smaller Python 3 at runtime
  than doesn't pull in some dev paths
* Build from Git
* Use autoreconfHook so that we don't have to patch ./configure scripts
  when we could just generate them correctly in the first place
* Drop patch that removes IBUS_* environment variables from ibus-setup
  wrapper: unsetting them breaks ibus-setup
* Add withWayland option (building with Wayland requires libxcbcommon
  package, previously wayland was included into buildInputs w/o
  libxcbcommon and respective configure flag)
* Add license (LGPLv2.1)

Relies on two other commits in this branch that add GI_TYPELIB_PATH
discovery and fix gconf Python scripts.

yegortimoshenko commented:
Well, it seems that `python3.withPackages` is just a shortcut to `python3.buildEnv.override`, which @edolstra doesn't like because it pulls in a lot of `dev` paths: https://github.com/NixOS/nixpkgs/issues/30265#issuecomment-335768590

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/with-packages.nix

Is there any way around this?

grahamc commented:
This seems to cause an evaluation error for me

yegortimoshenko commented:
@grahamc Strange, it didn't for me. Perhaps I've rebased incorrectly, hold on...

grahamc commented:
```
error: while evaluating anonymous function at /run/user/1000/nix-rebuild-amount-EstARSxV:13:12, called from /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/attrsets.nix:199:52:
while evaluating the attribute ‘x86_64-linux’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/attrsets.nix:185:41:
while evaluating anonymous function at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/top-level/release-lib.nix:66:6, called from /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/attrsets.nix:282:43:
while evaluating ‘hydraJob’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:157:14, called from /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/top-level/release-lib.nix:66:14:
while evaluating the attribute ‘drvPath’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:174:13:
while evaluating the attribute ‘drvPath’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:145:40:
while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘adapta-gtk-theme-3.91.2.147’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘gnome-shell-3.24.3’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘gnome-control-center-3.24.3’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating ‘getOutput’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/attrsets.nix:450:23, called from undefined position:
while evaluating anonymous function at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/stdenv/generic/make-derivation.nix:63:15, called from undefined position:
while evaluating ‘callPackageWith’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:113:35, called from /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/top-level/all-packages.nix:1474:10:
while evaluating ‘makeOverridable’ at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:72:24, called from /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:117:8:
anonymous function at /run/user/1000/nix-rebuild-amount-WfmnwnPI/pkgs/tools/inputmethods/ibus/default.nix:1:1 called without required argument ‘gconf’, at /run/user/1000/nix-rebuild-amount-WfmnwnPI/lib/customisation.nix:74:12

```

yegortimoshenko commented:
Oh... I should start testing packages via `nix-build -I nixpkgs=... -A ibus` instead of `nix-build -E "with import <nixpkgs> {}; callPackage ./. { ... }"`.

cstrahan commented:
Thanks @yegortimoshenko! For reference, there are some other tweaks that we might ultimately need, as described in my comments here: #30265

cstrahan commented:
I'm getting this error:

```
error: anonymous function at /home/cstrahan/src/nixpkgs/pkgs/tools/inputmethods/ibus/default.nix:1:1 called without required argument 'gconf', at /home/cstrahan/src/nixpkgs/lib/customisation.nix:74:12
```

This should fix that:

```patch
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 7383df20c0e..4ddc23072a8 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -1474,7 +1474,7 @@ with pkgs;
   skktools = callPackage ../tools/inputmethods/skk/skktools { };
 
   ibus = callPackage ../tools/inputmethods/ibus {
-    inherit (gnome3) dconf glib;
+    inherit (gnome3) gconf dconf glib;
   };
 
   ibus-qt = callPackage ../tools/inputmethods/ibus/ibus-qt.nix { };
```

yegortimoshenko commented:
Definitely, this isn't completely up to date. Also, set `enablePythonLibrary` to `false` for now, it tries to install Python libraries into read-only Python 3 directory.

cstrahan commented:
Now I'm getting this error:

```
 /nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/mkdir -p '/nix/store/vmzrpb6vxjir9g51anagwbcc1vs55qvx-ibus-1.5.17/share/dbus-1/services'
 /nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/install -c -m 644 org.freedesktop.IBus.service '/nix/store/vmzrpb6vxjir9g51anagwbcc1vs55qvx-ibus-1.5.17/share/dbus-1/services'
 /nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/mkdir -p '/nix/store/vmzrpb6vxjir9g51anagwbcc1vs55qvx-ibus-1.5.17/share/man/man1'
 /nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/install -c -m 644 ibus-daemon.1.gz '/nix/store/vmzrpb6vxjir9g51anagwbcc1vs55qvx-ibus-1.5.17/share/man/man1'
make[3]: Leaving directory '/build/source/bus'
make[2]: Leaving directory '/build/source/bus'
make[1]: Leaving directory '/build/source/bus'
Making install in ibus
make[1]: Entering directory '/build/source/ibus'
Making install in interface
make[2]: Entering directory '/build/source/ibus/interface'
make[3]: Entering directory '/build/source/ibus/interface'
make[3]: Nothing to be done for 'install-exec-am'.
 /nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/mkdir -p '/nix/store/sp876rk20728sv9di7nzj0bvrv9daa8z-python3-3.6.3-env/lib/python3.6/site-packages/ibus/interface'
/nix/store/8yd19hm3gq4vs4hwp6s2ydhzff5cnnv5-coreutils-8.28/bin/mkdir: cannot create directory '/nix/store/sp876rk20728sv9di7nzj0bvrv9daa8z-python3-3.6.3-env/lib/python3.6/site-packages/ibus': Permission denied
make[3]: *** [Makefile:487: install-ibus_interfacePYTHON] Error 1
make[3]: Leaving directory '/build/source/ibus/interface'
make[2]: *** [Makefile:583: install-am] Error 2
make[2]: Leaving directory '/build/source/ibus/interface'
make[1]: *** [Makefile:685: install-recursive] Error 1
make[1]: Leaving directory '/build/source/ibus'
make: *** [Makefile:682: install-recursive] Error 1
builder for '/nix/store/82aq5swxkid5kqfndg0sskx6a3r39bhd-ibus-1.5.17.drv' failed with exit code 2
cannot build derivation '/nix/store/x9ysrnqyi5mvwvg72fynffz1m2shpcx5-ibus-with-plugins-1.5.17.drv': 1 dependencies couldn't be built
error: build of '/nix/store/x9ysrnqyi5mvwvg72fynffz1m2shpcx5-ibus-with-plugins-1.5.17.drv' failed
```

yegortimoshenko commented:
See above :-)

cstrahan commented:
> See above :-)

Real life race conditions :)

cstrahan commented:
@yegortimoshenko 

This fixes the python lib, and everything builds now:

```patch
diff --git a/pkgs/tools/inputmethods/ibus/default.nix b/pkgs/tools/inputmethods/ibus/default.nix
index ba348229093..520d2934114 100644
--- a/pkgs/tools/inputmethods/ibus/default.nix
+++ b/pkgs/tools/inputmethods/ibus/default.nix
@@ -77,6 +77,9 @@ stdenv.mkDerivation rec {
     substituteInPlace data/dconf/Makefile.am --replace "dconf update" true
     echo \#!/bin/sh > data/dconf/make-dconf-override-db.sh
     cp ${gtk_doc}/share/gtk-doc/data/gtk-doc.make .
+    substituteInPlace configure.ac \
+      --replace 'pkgpython2dir="$python2dir/ibus"' \
+                'pkgpython2dir="'$out'/${python3.sitePackages}/ibus"'
   '';
 
   preAutoreconf = "touch ChangeLog";
```

yegortimoshenko commented:
Thanks :-) That's one way to fix it, another one that probably doesn't require patching Autoconf script is by redefining `PYTHON2_LIBDIR` to `$out`. Anyway, this isn't going to get merged until someone finds how to stop `python3.buildEnv` from pulling `dev` paths.

yegortimoshenko commented:
Without `stdenv.lib.closePropagation` in `pkgs/development/interpreters/python/wrapper.nix`, `ibus-setup` and `python3.withPackages` seem to work, and closure has significantly less `dev` paths in it:

```
$ nix-build ~/Projects/nixpkgs/ -A ibus
/nix/store/qlkr56zahbxny5f6y8aqb11420p3k3gi-ibus-1.5.17
$ nix-store --query --requisites result
/nix/store/mjx71lmnlf4psm9942djjcd8b56hyk8b-glibc-2.26-75
/nix/store/cqjq489nydhz27rjlrhcxx9bizwxr56j-attr-2.4.47
/nix/store/szf0cpzndm3ic2vh344sr024r19jdxx1-acl-2.2.52
/nix/store/0abay33zgwqgchdzzq6h88ac4963smwf-coreutils-8.28
/nix/store/43h1b33sickcv3d4gsm0ypsngza06x5l-pcre-8.41
/nix/store/nb8114h6xhkcf0l8jc42hr5fcqay908g-libffi-3.2.1
/nix/store/qz2lprii3jxy0wybggj4h3304ij7m56p-zlib-1.2.11
/nix/store/vly150v5s6l5y4pvh8lw2z09wq4syj1s-util-linux-2.31
/nix/store/0cgwsihhxi6vzix5vdchrvj8m4qzb3r1-glib-2.54.1
/nix/store/4ada72n7785wwazv42fhsnxjvilaa3aj-bash-4.4-p12
/nix/store/wfq4zjwgbjbxny0nv56ys3mc7i30x6ic-libIDL-0.8.14
/nix/store/0g71q1rgxmmv7bp44vyrrqn63prmasgd-ORBit2-2.14.19
/nix/store/ni3s4gas1c95aqkkr67pla6k66chkmb8-gcc-6.4.0-lib
/nix/store/dklc25dla8b5am9ywci4g4wmilzk50fc-gmp-6.1.2
/nix/store/4yv573zc7mv1m62yw84003r78d8swjaw-nettle-3.3
/nix/store/mlmwnxqd5p75lnwdmnw6i5vrh4fhfh2p-openssl-1.0.2m
/nix/store/hwb551a48jchz8kv3nxh629bk0pj8smk-libevent-2.1.8
/nix/store/jrasy4ivdfs0z117wra3l3ylwgkc5aaz-expat-2.2.5
/nix/store/0p4mn2xw4ahz1h7qsm20kfr1yc3jg92h-unbound-1.6.7-lib
/nix/store/5fcpyi3vlxbgllb4dnxrqydyfy2f7s8v-gdbm-1.13
/nix/store/az10kjba9f71m2kr39wcgqyv9klz347w-xz-5.2.3
/nix/store/gn56m9xpwk33xxrp8xrj2jr6d57jspic-ncurses-6.0-20170902
/nix/store/vbcyvh7rlbmpp85pssmcx4r8nn883dqa-bzip2-1.0.6.0.1
/nix/store/z90s4radpf90mdmmaw9ld33rs371rdgf-sqlite-3.21.0
/nix/store/zmjhihhfb9kxmb55zr3qw949mljfy2l1-readline-6.3p08
/nix/store/166s7l3yjqfc8dj5hfqjb09dbfvp1850-python3-3.6.3
/nix/store/1bkj4si57gq4hc9dmjkg5x41df04y3s2-python-recursive-pth-loader-1.0
/nix/store/1br4sydl06f16hwii9ryxxjqisw613lv-iso-codes-3.74
/nix/store/imyghld283zy7gkl0nc7kiabwv8s85c9-libICE-1.0.9
/nix/store/1gawvhhwpri8mdxjpfk5k24l39rbb2k7-libSM-1.2.2
/nix/store/1hmqll2jczwx0sksbd42qr08kvv8ys9g-zlib-1.2.11-dev
/nix/store/a83mwdh3xla231kjk1w475awzyk1yhfq-libpng-apng-1.6.34
/nix/store/hwqqym5ghdzfdzpc3g0za5w4m9pqnvs9-libjpeg-turbo-1.5.2
/nix/store/5gbphnrkqvm1rcffpy30ngpqx7fr7lml-libtiff-4.0.8
/nix/store/q87ni9psb36rh7as0xmw8r4i5sabg2iy-libXdmcp-1.1.2
/nix/store/x406s840qhnrckrphzqnzmkjacwqdinr-libXau-1.0.8
/nix/store/vmqni292w8yb9ds6bcadq7ym1n06pixs-libxcb-1.12
/nix/store/7kk1999bpav8lp097g671dxpjzcaxfqw-libX11-1.6.5
/nix/store/7wcyb7x9ll14xgb8h8s97zcd78j2j8ss-jasper-2.0.13
/nix/store/nmhcfv29nxng0hrdfag784n1zf60pmz3-gdk-pixbuf-2.36.7
/nix/store/1rq5xis1yg3ffqmggbmyg6mxk24r6yac-libnotify-0.7.7
/nix/store/4d2yvmg5mdxvn8lrlq7ifnkba7b0zx0n-libgpg-error-1.27
/nix/store/8q6k0cmwv81xz3gdfwyhj94k5fv8vs0h-libcap-2.25-lib
/nix/store/nrq4f4r3fljbsw7ky2ad52sbn18kjvxw-libgcrypt-1.8.1
/nix/store/r7ygpshlmfhjx9phh6h9qahj5zsykx6y-lz4-131
/nix/store/mdb9cz4hz554qqd6hvzsjv53awzgpqkj-systemd-234-lib
/nix/store/1zcqwpjfy78w1xaq799idxp4d36j0zm2-dbus-1.10.24-lib
/nix/store/410gjj0q962wbwy0ynchgx6ds1kwp1cv-libXfixes-5.0.2
/nix/store/4lgamb95j4jnlqiwxyf6nvhrx8r9vqkx-json-glib-1.4.2
/nix/store/56gzb238f6xgcr0i6rf8lyqb1ns4wml9-cracklib-2.9.6
/nix/store/5abak7azq3inwwvss2ldkpxxs9rh6r5v-libtasn1-4.12
/nix/store/56qlrnhpkr1qgg5gascvdvlskdpia82z-p11-kit-0.23.7
/nix/store/7vqbmwjv8acb3afqvd9bip13xh6f12ps-glibc-2.26-75-bin
/nix/store/ilhkswnnncfz49lrzkfd9sglp3amik06-linux-headers-4.4.10
/nix/store/xd2mlzvlgmskbyyb33lz2vdq15k74xpr-glibc-2.26-75-dev
/nix/store/796cdc13vzh79c101j9x34llyxk0f93w-glibc-iconv-2.26-75
/nix/store/d5lcw25l1xkdaj034hm2gxn64acb50rk-libelf-0.8.13
/nix/store/xzn1wijp8kk9r6nhcjrhlamdz6i9caar-db-5.3.28
/nix/store/f6pmgza1hha574hm76p98mc2s6gsn3y8-python-2.7.14
/nix/store/ivm3ba9dzw26yggkxlfkxlbly67f35jk-libffi-3.2.1-dev
/nix/store/5bglj2xcnyh54c6jp13mcyrdg2cgc0nq-glib-2.54.1-dev
/nix/store/685105cs828ci90pj112y0dnh7zx1ha4-graphite2-1.3.6
/nix/store/6b12yg3y7dbak7k99fxkz3n0fzaz0r0g-freetype-2.7.1
/nix/store/wpfsqydmgwxbfh3rkslhah2c774c7gmw-nspr-4.16
/nix/store/75ixdg30k57icaxgi5h8v3h9k3l8r8l6-spidermonkey-17.0.0-lib
/nix/store/lwqfv7k2c227d29xjb81dk6lq8h0h0ry-libXext-1.3.3
/nix/store/8nzbr0k8dwvnjac2wili08im3il1rl05-libXxf86vm-1.1.4
/nix/store/8gn2b5vvlazg608cj1y5l4igp9rckmnq-dejavu-fonts-minimal-2.37
/nix/store/aaj5jf73ac4jwv885djxrs4drqkv0jhx-fontconfig-2.12.1-lib
/nix/store/c9pv9kl83134gspxmy6dbq41r1kzksbh-libxshmfence-1.2
/nix/store/imlh65g6gxcj4b9fni46pvzwmpr9czx2-libxml2-2.9.5
/nix/store/ck57nbay428qi5dx3r0fr2paqnxh9jci-wayland-1.12.0
/nix/store/hlhvvc6ghc8l0bwk6rs8grzs6rxdm4fn-libXrender-0.9.10
/nix/store/rbc04km5yfbjwrsn9mzsw1bybmg43s9v-libXdamage-1.1.4
/nix/store/xdqb5gzl5gr9vdi1xshnk53yligpals2-libpciaccess-0.13.5
/nix/store/vvng84xliafz3iwzxbm7cf17niwp00zp-libdrm-2.4.84
/nix/store/q353py5mrssppbfzpamm2n4n79w4l984-mesa-noglu-17.2.4
/nix/store/swi9n1azsvv36bsbpm98pn87zlsm6vi6-pixman-0.34.0
/nix/store/gy1cj8n212pgx400f34615z1cdy1krji-cairo-1.14.10
/nix/store/yiixwpqlnxxisx64q9jb043049751qbp-gobject-introspection-1.54.1
/nix/store/7mdkm7j6kn18zsadkv6g0j2cpwxg8v5r-python3.6-pygobject-3.24.1
/nix/store/8hc2a9khyzn2w2l6vlvqsmh3pmgkky0y-linux-pam-1.2.1
/nix/store/hhn4n31vs8rwj334wyxczf5n3rp4fb4z-libXft-2.3.2
/nix/store/ryn59z81fxiarj12zjxmv1b4j9055hf8-harfbuzz-1.5.1
/nix/store/fzr2i9n9xjh78dajxjj2kcbp1g26qzmw-pango-1.40.12
/nix/store/lj64zydvbdq4gv0bj3bhl3aizas0a8p9-libcroco-0.6.12
/nix/store/8m0pfs9n6k7jbc4jk521czw5blrxb9mr-librsvg-2.40.19
/nix/store/9kv1jgyrpg5jhgwk3i02ryhp5snh8xdy-dbus-glib-0.108
/nix/store/kqj9qj2rgh9jgv006qnpsghbgnjvqs1i-libXi-1.7.9
/nix/store/xvpby6pb0r8hbnjgdglinwzb4ilkp9wy-recordproto-1.14.2
/nix/store/df6c4ard0klpyz6aa37lmc1fbjq4xf0i-libXtst-1.2.3
/nix/store/9s7gl0ndmgwlv8sk9dxsy199yqm46l1v-at-spi2-core-2.26.0
/nix/store/j5g8jhqphdv5pzvsvgrvnygaddas966d-libXinerama-1.1.3
/nix/store/sfkaq45x6isibn64zilng19jagnn15sh-libdaemon-0.14
/nix/store/fkxb1iwy9iz7p5srqdhnaa8rhhhi860g-avahi-0.7
/nix/store/brrqxa283wzcnbm2dlg39frr8475cr8x-libunistring-0.9.7
/nix/store/j4rjrn15yrcybw1yj4iqmqym7srl9cw0-dns-root-data-2017-10-24
/nix/store/v7h2jyc5c8qk3ypli4lbvy66pg5031ra-libidn-1.33
/nix/store/yfnm9yla579w16gwf3biy0ipj0s8v2z1-gnutls-3.5.15
/nix/store/l4k0whw94b4666ws6hxbfhp6q1n2d9dk-cups-2.2.2-lib
/nix/store/m7201cf31zgb2c295irn1a82la83dxf0-atk-2.26.0
/nix/store/mfl11v2qif9l1ib3ld1c275g63wicx2f-libXrandr-1.5.1
/nix/store/rpd9r35pcvyqy86mnln9ig7cmsvwjhj8-libXcursor-1.1.14
/nix/store/xr7hldw8wafls0vl2y68d62ynw0qafk1-libXcomposite-0.4.4
/nix/store/albn8pji19fa6rpa6r1kxydv28iddw0j-gtk+-2.24.31
/nix/store/aq9n99qlrd4mv5xzssgkwcx6vb7ywlvy-python3-3.6.3-env
/nix/store/giav578rb9zgrl1firrkh6p2md2gnf00-at-spi2-atk-2.26.0
/nix/store/shj01cjd8ra57wb9p17n043dmvn1crbm-polkit-0.113
/nix/store/mfkqcsdja6a5m5nxqq2m50nks79ls3g1-gconf-3.2.6
/nix/store/r1ndb8czlf6ybp7bp0dmjqfvlgg10g66-epoxy-1.3.1
/nix/store/wg9hblgn2swhg79qn9qa3xq0v735y9d4-xkeyboard-config-2.22
/nix/store/ry35zgalhin9gfxrdb1l2rdfsqgjpc85-libxkbcommon-0.7.2
/nix/store/shq5c3qh1mdmb3bnfrhnisz0bsq5k1xq-gtk+3-3.22.24
/nix/store/v9izwgkp45n9m7pa04gq72fg2a5vs3i1-dconf-0.26.0-lib
/nix/store/qlkr56zahbxny5f6y8aqb11420p3k3gi-ibus-1.5.17
```

yegortimoshenko commented:
Got `dev` paths to zero by explicitly specifying output of `glib` that we want to propogate:

```diff
- propagatedBuildInputs = [ glib ];
+ propagatedBuildInputs = [ glib.out ];
```

FRidh commented on pkgs/development/interpreters/python/wrapper.nix:
```diff
@@ -8,7 +8,7 @@
   recursivePthLoader = import ../../python-modules/recursive-pth-loader/default.nix { stdenv = stdenv; python = python; };
   env = let
-    paths = stdenv.lib.closePropagation (extraLibs ++ [ python recursivePthLoader ] ) ;
+    paths = extraLibs ++ [ python recursivePthLoader ];
```
And how are the listed `extraLibs` Python packages now supposed to find their dependencies? This can't go in. 

I don't know if you intended to but changes like these require a separate pull request as this is a major change.

yegortimoshenko replied:
Sure, this PR is just trying to discover how to get `ibus` fixed in mergeable fashion. `closePropagation` is deprecated anyway and should be replaced with something else. Could you explain why the environment has to have complete closure in it?

FRidh replied:
We indeed need to get rid of the use of `closePropagation`. The issue we have is that we don't really mark packages providing Python libraries (read: they have `site-packages/`) as such, and we do not use such marking/attribute to determine dependencies. Therefore, a blunt approach is taken by just recursing through all `propagatedBuildInputs`. A possible solution would be https://github.com/NixOS/nixpkgs/pull/30606/files.

Due to how wrappers are created also more than needed remains inside closures.

yegortimoshenko replied:
Well, I think we could filter out only those packages that have `site-packages/` at script runtime, and build environment with only those. That might require forking, or changing `buildEnv` Perl script, but doable. Is it a viable option?

FRidh replied:
Not really. You have to track which Python packages belong together. E.g., imagine you have a new Python env that has in it a program that requires another Python env. You need to be able to distinguish those two. This is mostly an issue when you include in your env an application written in Python. There you may want to hide its `site-packages`.

FRidh commented on pkgs/tools/inputmethods/ibus/default.nix:
```diff
@@ -1,129 +1,142 @@
-        -i "setup/ibus-setup.in"
+  patchPhase = ''
+    # https://github.com/NixOS/nixpkgs/issues/30265#issuecomment-335768590
+    substituteInPlace setup/ibus-setup.in --subst-var-by PYTHON ${python3Runtime}/bin/python
```
`$(python3Runtime.interpreter)`

yegortimoshenko replied:
Thanks!

FRidh commented on pkgs/tools/inputmethods/ibus/default.nix:
```diff
@@ -1,129 +1,142 @@
+    isocodes
+    json_glib
+    libnotify
+    python3Runtime
```
Both here and in `nativeBuildInputs` is a Python env. What is the reason for that, and how is that working? Say patchShebangs patches Python shebangs, to which env will it rewrite them?

yegortimoshenko replied:
`python3BuildEnv` pulls in `glib.dev` (that we only need at build time), and `python3Runtime` doesn't. There are no shebangs to patch per se, which is why this works. The only occurence of Python is patched above:

```
substituteInPlace setup/ibus-setup.in --subst-var-by PYTHON ${python3Runtime}/bin/python
```

yegortimoshenko replied:
Dropped `python3Runtime` from the list to avoid confusion.

yegortimoshenko commented:
@FRidh Please check https://github.com/NixOS/nixpkgs/pull/31094/commits/60b3cd5f857d327f43c1c9519c5f712955a6820e. It should fix closure size problems for all `python3.withPackages` uses, not just this one. And yes, I might open a separate PR if necessary.

yegortimoshenko commented on pkgs/development/interpreters/python/wrapper.nix:
```diff
@@ -1,59 +1,76 @@
-          cd "$path/bin"
-          for prg in *; do
-            if [ -f "$prg" ]; then
-              rm -f "$out/bin/$prg"
```
This is the only functional change between the previous `python3.buildEnv` and this one: if `$path/bin/$prg` is a non-executable file (which is arguably a packaging mistake), `$out/bin/$prg` is not deleted anymore.

yegortimoshenko commented:
I've refactored `python3.buildEnv` and added a `closePropagation` wrapper that sets `outputUnspecified = true;` to all derivations it produces, and copied some of `buildEnv` logic to pick outputs to use in `postBuild`. Now it pulls in one `dev` path, `glibc.dev`:


```
$ nix-build ~/Projects/nixpkgs/ -A ibus
/nix/store/9n9nls1z5fwcl6y0rv0i735win0v0ry8-ibus-1.5.17
$ nix-store -qR result | grep dev
/nix/store/xd2mlzvlgmskbyyb33lz2vdq15k74xpr-glibc-2.26-75-dev
```

FRidh commented:
@yegortimoshenko can you open a separate PR with the `python.buildEnv` change? Can you also test against https://github.com/NixOS/nixpkgs/pull/30606. There are still some issues with it, but it should solve the `dev` issue.

yegortimoshenko commented:
@FRidh Definitely, separating it out.

FRidh commented:
I've pushed the changes in #30606 to staging, so that should solve your issue.

yegortimoshenko commented:
Thanks! I'll rebase this branch on top of staging.

FRidh commented:
Note the change is in master now as well.

yegortimoshenko commented:
This is ready to merge! Now with zero `dev` paths, thanks to @FRidh and #30606.

yegortimoshenko commented:
This now targets `staging` (I'd like to make sure that `wrapGAppsHook` changes haven't broken anything). Could someone please review this?

jtojnar commented on pkgs/tools/inputmethods/ibus/default.nix:
```diff
@@ -1,129 +1,141 @@
-    dbus libnotify gobjectIntrospection wayland
+  nativeBuildInputs = [
+    autoreconfHook
+    gconf
```
Why are you adding `gconf` when it is disabled by `configureFlags`?

yegortimoshenko replied:
It's still needed by the build system, unfortunately. It isn't pulled into the closure, though. Disabling `gconf` removes it from the closure, but not from the build.

yegortimoshenko replied:
```
<jtojnar> yegortimoshenko: any idea why gconf is suddenly required?  [04:33]
<yegortimoshenko> jtojnar: maybe for some tools?
<jtojnar> seems to be called from here
	  https://github.com/ibus/ibus/blame/master/data/dconf/Makefile.am#L54
								        [04:35]
<yegortimoshenko> right  [04:36]
<yegortimoshenko> $ find /nix/store -name gsettings-schema-convert
<yegortimoshenko>
		 /nix/store/78a9m0l4x60b0l70nl9q5w7v226dwb1a-gconf-3.2.6/bin/gsettings-schema-convert
```

jtojnar replied:
Apparently, the GSettings schema generated from the GConf one is listed in `EXTRA_DIST` so it is included in the dist tarball. Since we now no longer use dist tarball, we need to generate it ourselves.

https://github.com/ibus/ibus/pull/1813

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook.sh:
```diff
@@ -35,6 +35,10 @@ wrapGAppsHook() {
   done
 
+  if [ -d "$prefix/lib/girepository-1.0" ]; then
+    gappsWrapperArgs+=(--prefix GI_TYPELIB_PATH : "$prefix/lib/girepository-1.0")
```
This is supposed to be populated via `gobjectIntrospection` setup hook and added in the loop above.

yegortimoshenko replied:
When I drop this patch and build IBus, I get:

```
$ result/bin/ibus-setup
Traceback (most recent call last):
  File "/nix/store/rlz1ssvicswx87v2ry7pfff88chw4kic-ibus-1.5.17/share/ibus/setup/main.py", line 38, in <module>
    gi_require_version('IBus', '1.0')
  File "/nix/store/nmr66mmvhhz9qyshzv9ap72g4rjfbw5j-python3-3.6.3-env/lib/python3.6/site-packages/gi/__init__.py", line 130, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace IBus not available
```

I'll rebase it on top of current master and check if it works now, but at least it didn't back when I started this PR.

yegortimoshenko replied:
@jtojnar It doesn't work without this patch, even when rebased on top of `master`. Should I open a separate issue? Or is that an adequate fix?

yegortimoshenko replied:
I've made the proper fix, see dd0db4e016ccfce8eb6b2ca3f04be71f0babc5ab.

jtojnar commented on pkgs/tools/inputmethods/ibus/default.nix:
```diff
@@ -1,129 +1,141 @@
-        -e "/export IBUS_LOCALEDIR/ s/^.*$//" \
-        -e "/export IBUS_PREFIX/ s/^.*$//" \
-        -i "setup/ibus-setup.in"
+  patchPhase = ''
```
This should probably still be `postPatch` in case we want to add `patches`.

yegortimoshenko replied:
True, thanks.

lukateras committed:
gobjectIntrospection: add giDiscoverSelf preFixupPhase hook

Some derivations need to discover their own typelibs. giDiscoverSelf
runs as a preFixupPhase and populates GI_TYPELIB_PATH with an entry
pointing to $prefix/lib/girepository-1.0, if it exists.

yegortimoshenko commented:
@GrahamcOfBorg build

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

yegortimoshenko commented:
@GrahamcOfBorg build ibus

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
>

GrahamcOfBorg reviewed (commented):
Success for system: aarch64-linux

```
no configure script, doing nothing
building
no Makefile, doing nothing
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/fr3xmhr4mdh0xv5qwr8shq00133h422f-python3.6-bootstrapped-pip-9.0.1
stripping (with flags -S) in /nix/store/fr3xmhr4mdh0xv5qwr8shq00133h422f-python3.6-bootstrapped-pip-9.0.1/lib  /nix/store/fr3xmhr4mdh0xv5qwr8shq00133h422f-python3.6-bootstrapped-pip-9.0.1/bin
patching script interpreter paths in /nix/store/fr3xmhr4mdh0xv5qwr8shq00133h422f-python3.6-bootstrapped-pip-9.0.1
checking for references to /build in /nix/store/fr3xmhr4mdh0xv5qwr8shq00133h422f-python3.6-bootstrapped-pip-9.0.1...
/nix/store/34hg7sdvws10c4r9bglbnadvxspww2dj-ibus-1.5.17
```

GrahamcOfBorg reviewed (commented):
Failure for system: x86_64-linux

```
  CC       window-default.o
  CCLD     window-default
  CC       plugman.o
  CCLD     plugman
  CC       sunny.o
  CCLD     sunny
building of ‘/nix/store/2ggnk3akh30z0bml1x0g0fsh5qwibazq-gtk+3-3.22.26.drv’ timed out after 3600 seconds
cannot build derivation ‘/nix/store/ps9klyxg26w95y3ap06jqc1bgvvvqydv-hook.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/3pp9lwaai9x8hk2lpbz8v416z1nwimmn-ibus-1.5.17.drv’: 2 dependencies couldn't be built
error: build of ‘/nix/store/3pp9lwaai9x8hk2lpbz8v416z1nwimmn-ibus-1.5.17.drv’ failed
```

GrahamcOfBorg reviewed (commented):
Failure for system: x86_64-linux

```
libtool: warning: relinking 'im-inuktitut.la'
libtool: install: (cd /tmp/nix-build-gtk+3-3.22.26.drv-0/gtk+-3.22.26/modules/input; /nix/store/65l6hr8snf4v823f974k97jc65i7bhvf-bash-4.4-p12/bin/bash "/tmp/nix-build-gtk+3-3.22.26.drv-0/gtk+-3.22.26/libtool"  --silent --tag CC --mode=relink gcc -g -O2 -Wall -rpath /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules -avoid-version -module -o im-inuktitut.la -rpath /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules iminuktitut.lo ../../gtk/libgtk-3.la -L/nix/store/ciz98qjymi65iaq535nylgi36mx9m6jl-glib-2.54.2/lib -L/nix/store/vgxcfqdr0bg7208lx1zvqgk1552aj2jw-libxkbcommon-0.7.2/lib -L/nix/store/5xa736rky9qh6f7q6bmrifi3dv69frkc-epoxy-1.3.1/lib -L/nix/store/xsc1vqqs5d72f1pjfds5sdf6qzjgqj7b-cairo-1.14.10/lib -L/nix/store/93n4kksb0bkmx8j980pg1yxin4xfmzdn-fontconfig-2.12.1-lib/lib -L/nix/store/gxyg8jmcfw3khyj1qi1sha15ysd8yvig-freetype-2.7.1/lib -L/nix/store/x4cqrnzcvcc1lqvz41cp0dxqbnvsmhfk-libX11-1.6.5/lib -L/nix/store/bwrmg4vx7734hzyg41raaix5w12jnbzc-mesa-noglu-17.2.7/lib -L/nix/store/ihc09b12dnl7a596shvanrw34bvnkgkm-libXdamage-1.1.4/lib -L/nix/store/8jhjp4v16arm59cww1s9aha6hbjz7cwr-libXfixes-5.0.2/lib -L/nix/store/n0ng6gd5rcv5yz9w9rpdhpcwpnbmgq0s-pango-1.40.14/lib -L/nix/store/3z0rrpdfsiha2qjrhcbw0y3di0cv9nkz-gdk-pixbuf-2.36.7/lib -L/nix/store/djnbk5pmnb0zj8h2hxlmkyz4z5dck5gl-atk-2.26.1/lib -L/nix/store/vhhpp8hq48w7d1m8sdzhnzw1wyrn14zj-at-spi2-atk-2.26.1/lib -L/nix/store/mk0bgbf7y6ixygmr7yrkr0287bijmhdp-libXcomposite-0.4.4/lib -L/nix/store/jkgrv5ga89j8z4wfh9c3qk5xm01qzcby-libXi-1.7.9/lib -L/nix/store/8y125i89fkkx7x84fdgx2ad1cf3my4ml-wayland-1.14.0/lib -lpangocairo-1.0 -lX11 -lXi -lXcomposite -lXdamage -lXfixes -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -latk-1.0 -latk-bridge-2.0 -lxkbcommon -lwayland-cursor -lwayland-egl -lwayland-client -lepoxy -lpangoft2-1.0 -lpango-1.0 -lfontconfig -lfreetype -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lm )
libtool: install: /nix/store/hw5a3mifkrzd8y0pxs7nzzr4yscg08mw-coreutils-8.28/bin/install -c .libs/im-inuktitut.soT /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules/im-inuktitut.so
libtool: install: /nix/store/hw5a3mifkrzd8y0pxs7nzzr4yscg08mw-coreutils-8.28/bin/install -c .libs/im-inuktitut.lai /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules/im-inuktitut.la
libtool: warning: relinking 'im-ipa.la'
libtool: install: (cd /tmp/nix-build-gtk+3-3.22.26.drv-0/gtk+-3.22.26/modules/input; /nix/store/65l6hr8snf4v823f974k97jc65i7bhvf-bash-4.4-p12/bin/bash "/tmp/nix-build-gtk+3-3.22.26.drv-0/gtk+-3.22.26/libtool"  --silent --tag CC --mode=relink gcc -g -O2 -Wall -rpath /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules -avoid-version -module -o im-ipa.la -rpath /nix/store/32w8kdpy6ykj23q95vkdv50zxxiclv6g-gtk+3-3.22.26/lib/gtk-3.0/3.0.0/immodules imipa.lo ../../gtk/libgtk-3.la -L/nix/store/ciz98qjymi65iaq535nylgi36mx9m6jl-glib-2.54.2/lib -L/nix/store/vgxcfqdr0bg7208lx1zvqgk1552aj2jw-libxkbcommon-0.7.2/lib -L/nix/store/5xa736rky9qh6f7q6bmrifi3dv69frkc-epoxy-1.3.1/lib -L/nix/store/xsc1vqqs5d72f1pjfds5sdf6qzjgqj7b-cairo-1.14.10/lib -L/nix/store/93n4kksb0bkmx8j980pg1yxin4xfmzdn-fontconfig-2.12.1-lib/lib -L/nix/store/gxyg8jmcfw3khyj1qi1sha15ysd8yvig-freetype-2.7.1/lib -L/nix/store/x4cqrnzcvcc1lqvz41cp0dxqbnvsmhfk-libX11-1.6.5/lib -L/nix/store/bwrmg4vx7734hzyg41raaix5w12jnbzc-mesa-noglu-17.2.7/lib -L/nix/store/ihc09b12dnl7a596shvanrw34bvnkgkm-libXdamage-1.1.4/lib -L/nix/store/8jhjp4v16arm59cww1s9aha6hbjz7cwr-libXfixes-5.0.2/lib -L/nix/store/n0ng6gd5rcv5yz9w9rpdhpcwpnbmgq0s-pango-1.40.14/lib -L/nix/store/3z0rrpdfsiha2qjrhcbw0y3di0cv9nkz-gdk-pixbuf-2.36.7/lib -L/nix/store/djnbk5pmnb0zj8h2hxlmkyz4z5dck5gl-atk-2.26.1/lib -L/nix/store/vhhpp8hq48w7d1m8sdzhnzw1wyrn14zj-at-spi2-atk-2.26.1/lib -L/nix/store/mk0bgbf7y6ixygmr7yrkr0287bijmhdp-libXcomposite-0.4.4/lib -L/nix/store/jkgrv5ga89j8z4wfh9c3qk5xm01qzcby-libXi-1.7.9/lib -L/nix/store/8y125i89fkkx7x84fdgx2ad1cf3my4ml-wayland-1.14.0/lib -lpangocairo-1.0 -lX11 -lXi -lXcomposite -lXdamage -lXfixes -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -latk-1.0 -latk-bridge-2.0 -lxkbcommon -lwayland-cursor -lwayland-egl -lwayland-client -lepoxy -lpangoft2-1.0 -lpango-1.0 -lfontconfig -lfreetype -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lm )
building of ‘/nix/store/2ggnk3akh30z0bml1x0g0fsh5qwibazq-gtk+3-3.22.26.drv’ timed out after 3600 seconds
cannot build derivation ‘/nix/store/ps9klyxg26w95y3ap06jqc1bgvvvqydv-hook.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/3pp9lwaai9x8hk2lpbz8v416z1nwimmn-ibus-1.5.17.drv’: 2 dependencies couldn't be built
error: build of ‘/nix/store/3pp9lwaai9x8hk2lpbz8v416z1nwimmn-ibus-1.5.17.drv’ failed
```

yegortimoshenko commented:
@orivej `giDiscoverSelf` hook now finally works and iterates through all outputs.

This should be ready to merge.

orivej commented:
I'm in the process of switching to the system with this PR, will merge after testing.

orivej commented:
After rebasing this on the current master (for testing), `ibus-engines.m17n` does not build with:
```
checking for
    ibus-1.0 >= 1.4
... no
configure: error: Package requirements (
    ibus-1.0 >= 1.4
) were not met:

Package 'gobject-2.0', required by 'ibus-1.0', not found
```

vcunat commented:
Adding `glib` to its `buildInputs` fixes the build. (And other engines do build as they are.)  This is caused by `ibus` propagating only `glib.out` instead of the usual `glib`, so that the pkg-config file is missing.

yegortimoshenko commented:
@orivej, @vcunat: Thanks a lot!

laMudri commented:
I've been trying this out on nixos-unstable since yesterday, and it seems to work as well as before.

yegortimoshenko commented:
@laMudri: Thanks for testing! I have also tried it out, and changing `glib.out` to just `glib` seemed to have solved the problem. I am merging this into staging.
